src/drivers/intel/cavs/mn.c

```diff
@@ -142,73 +171,190 @@ static bool find_mn(uint32_t freq, uint32_t bclk,
 	return true;
 }
 
-int mn_set_bclk(uint32_t dai_index, uint32_t bclk_rate,
-		uint32_t *out_scr_div, bool *out_need_ecs)
+/**
+ * \brief Finds index of clock valid for given BCLK rate.
+ *	  Clock that can use just SCR is preferred.
+ *	  M/N other than 1/1 is used only if there are no other possibilities.
+ * \param[in] bclk Bit clock frequency.
+ * \param[out] scr_div SCR divisor.
+ * \param[out] m M value of M/N divider.
+ * \param[out] n N value of M/N divider.
+ * \return index of suitable clock if could find it, -1 otherwise.
+ */
+static int find_bclk_source(uint32_t bclk,
+			    uint32_t *scr_div, uint32_t *m, uint32_t *n)
 {
-	uint32_t i2s_m = 1;
-	uint32_t i2s_n = 1;
-	uint32_t mdivc;
 	int i;
-	int clk_index = -1;
-	int ret = 0;
 
-	*out_need_ecs = false;
+	/* searching the smallest possible bclk source */
+	for (i = 0; i <= MAX_SSP_FREQ_INDEX; i++)
+		if (ssp_freq[i].freq % bclk == 0) {
+			*scr_div = ssp_freq[i].freq / bclk;
+			return i;
+		}
 
-	spin_lock(mn_lock);
-	mdivc = mn_reg_read(0x0);
+	/* check if we can get target BCLK with M/N */
+	for (i = 0; i <= MAX_SSP_FREQ_INDEX; i++)
+		if (find_mn(ssp_freq[i].freq, bclk,
+			    scr_div, m, n))
+			return i;
 
-	/* searching the smallest possible bclk source */
-	for (i = MAX_SSP_FREQ_INDEX; i >= 0; i--) {
-		if (bclk_rate > ssp_freq[i].freq)
-			break;
+	return -1;
```

Comment:
-EINVAL?

Reply:
`EINVAL` is for invalid argument, here's result is `not found`. Usually functions that look for index return 0-based index or -1 on `not found`. `-1` here is not error code it's result indicating that value wasn't find in table.